### ISSUE TITLE
types in extraction

### ISSUE THREAD

**Issue by aa755msr** [FStarLang/FStar#306]
It should at least be possible to have terms decorated with types. This can prevent some mistakes, both in extraction, and while manually implementing some functions in the extract.

We had some issues with OCaml refusing to generalize sometimes when terms were decorated with types. IIRC, those issues were gone when we manually added some type lambdas in OCaml. Is there some documentation about such type lambdas in OCaml?

**Comment by msprotz** [comment]
Fixing this will probably fix #323

**Comment by mtzguido** [comment]
Every top-level definition is now annotated in extracted code. Example:
```
$ cat examples/hello/hello.fst
module Hello

open FStar.IO

let main = print_string "Hello F*!\n"

let foo (x:(x:int & (y:int{y > x}) & int)) :int = match x with
  | Mkdtuple3 x _ _ -> x

$ cat Hello.ml
open Prims
let main: Prims.unit = FStar_IO.print_string "Hello F*!\n"
let foo:
  (Prims.int,Prims.int,Prims.int) FStar_Pervasives.dtuple3 -> Prims.int =
  fun x  ->
    match x with | FStar_Pervasives.Mkdtuple3 (x1,uu____72,uu____73) -> x1
```
Closing, but please reopen if this is not enough!
